modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp

```diff
@@ -269,31 +282,52 @@ template<typename U> struct get_out<cv::GOpaque<U>>
     }
 };
 
+template<typename, typename>
+struct OCVSetupHelper;
+
+template<typename Impl, typename... Ins>
+struct OCVSetupHelper<Impl, std::tuple<Ins...>>
+{
+    template<int... IIs>
+    static void setup_impl(GMetaArgs metaArgs, GArgs args, GArg &state, detail::Seq<IIs...>)
+    {
+        // unique_ptr <-> shared_ptr conversion ?
+        // To check: Conversion is possible only if the state which should be passed to
+        // 'setup' user callback isn't required to have previous value
+        std::shared_ptr<typename Impl::State> stPtr;
+        Impl::setup(detail::get_in_meta<Ins>(metaArgs, args, IIs)...,
+                    stPtr);
+        state = GArg(stPtr);
+    }
+
+    static void setup(GMetaArgs metaArgs, GArgs args, GArg& state)
+    {
+        setup_impl(metaArgs, args, state,
+                   typename detail::MkSeq<sizeof...(Ins)>::type());
+    }
+};
+
 template<typename, typename, typename>
 struct OCVCallHelper;
 
 // FIXME: probably can be simplified with std::apply or analogue.
 template<typename Impl, typename... Ins, typename... Outs>
-struct OCVCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...> >
+struct OCVCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...>>
 {
-    template<typename... Inputs>
-    struct call_and_postprocess
+    template<typename... Outputs>
+    static void postprocess(Outputs&&... outs)
     {
-        template<typename... Outputs>
-        static void call(Inputs&&... ins, Outputs&&... outs)
-        {
-            //not using a std::forward on outs is deliberate in order to
-            //cause compilation error, by trying to bind rvalue references to lvalue references
-            Impl::run(std::forward<Inputs>(ins)..., outs...);
-            postprocess(outs...);
-        }
+        ::cv::detail::postprocess(outs...);
+    }
```

Comment:
What's the reason to have this?

Reply:
The reason is in that that cv::detail::postprocess accepts arguments by reference, so arguments should be lvalues